### PR TITLE
feat(auth): bootstrapping process for first token

### PR DESCRIPTION
**Pull request by GeorgeMac**
Supports: #779 

Fixes: FLI-37
Fixes: FLI-38

Currently based on #1110, but still ready for review. I will fix up onto `authentication` when that PR is merged.

The primary purpose of this PR is to introduce a bootstrapping process, when the `token` authentication method is enabled.
This ensures at-least one token is present in the backing storage layer.
When created for the first time, the token is printed via an `INFO` log line with the message `"access token created"`.

There were a number of other structural / refactoring changes along the way to validate this behaviour appropriately.
This PR:

1. Introduces a new `storage/testing` package.

It currently contains a harness for any implementation of `storage.AuthenticationStore`.

The harness:

- Attempts to create 100 authentications.
- Re-retrieve them using the tokens produced when creating them.
- List all of them again in ascending and then descending order (+ compare with original set).

2. Implements `ListAuthentications` for both `auth/sql` and `auth/memory`.
3. Integrates the harness from (1) into both `auth/sql` and `auth/memory`.
4. Adds a test to `auth/sql` to ensure listing by auth method is supported.
5. Defines `internal/storage/auth/bootstrap.go` which contains a function to bootstrap an initial token if one is not present in the provided store.
6. Installs bootstrapping and token auth into `main.go`.
7. Updates `internal/config` to reflect discussion has with @markphelps around config needs.

### CODE REVIEW

**Review by markphelps — COMMENTED** [review]
couple questions/comments, but overall looking great!

File: cmd/flipt/main.go [hunk 1]
```diff
@@ -474,19 +476,35 @@ func run(ctx context.Context, logger *zap.Logger) error {
...
 			grpc_prometheus.UnaryServerInterceptor,
 			otelgrpc.UnaryServerInterceptor(),
+		}
+
+		// authentication interceptor
+		var authenticationStore storage.AuthenticationStore
+
+		// only enable enforcement middleware if authentication required
+		if cfg.Authentication.Required {
+			authenticationStore = authsql.NewStore(driver, sql.BuilderFor(db, driver), logger)
```
[the remark this answers is not in the document]
Comment (GeorgeMac) [reply cmd/flipt/main.go:495]: Yes, I think you've eluded to a subtle bug here.

We don't need a store if both `required` is `false` and none of the authentication methods are enabled. I will fix that.

In this case, we would panic if the token store was enabled but authentication was not required.

File: internal/storage/storage.go [hunk 2]
```diff
@@ -46,6 +54,18 @@ type QueryParams struct {
 	Order     Order // not exposed to the user yet
 }
 
+func (q *QueryParams) Validate() error {
+	if q.Limit == 0 {
+		q.Limit = DefaultListLimit
+	}
+
+	if q.Limit > MaxListLimit {
+		return fmt.Errorf("maximum page size limit (%d) exceeded: %d", MaxListLimit, q.Limit)
```
[the remark this answers is not in the document]
Comment (GeorgeMac) [reply internal/storage/storage.go:63]: Yeah, only if we introduced `Validate()` call in the existing packages. Which I haven't done in this PR.
If we're worried about inadvertently introducing that later, I could move this into the auth package?

[the remark this answers is not in the document]
Comment (GeorgeMac) [reply internal/storage/storage.go:63]: r.e. should we just snap to the max limit? Good question.

At-first, my feelings were that is implicit behaviour that might be surprising to an end user.
I request 1000 and get back 100.

However, I can see how this might be more jarring; I request 1000, get back a 4XX error (which I think I haven't actually done the work to make it a validation error—I would do that if we keep down this path).
I have to read the error and adjust my code to match limits.

I think if pagination was the way it used to be, I would be less inclined to snap the page to max.
As this could lead to mistakes when implementing pagination (end-user _must_ rely on response length to set next offset; they might mistakenly use `old_offset + limit` when requesting next page).
Given pagination is driven by page token, there is less meaning to the number of results in the response.
If you want more or all, you just keep following the page tokens until they end.
You can request 1000, but you will only get the max page size in any given response. Not an error.

TL;DR I will do as suggested and snap limit to max 👍

[the remark this answers is not in the document]
Comment (GeorgeMac) [reply internal/storage/storage.go]: See commit 5f94153e where I renamed `Validate()` to `Normalize()` and drop the error for this change 👍

**Comment by markphelps** [comment]
looks good to me! Couple comments, non-blocking ✅ 

TY!!

File: internal/storage/storage.go [hunk 3]
```diff
@@ -9,6 +9,14 @@ import (
 	"google.golang.org/protobuf/types/known/timestamppb"
 )
 
+const (
+	// DefaultListLimit is the default limit applied to any list operation page size when one is not provided.
+	DefaultListLimit uint64 = 10
```
[the remark this answers is not in the document]
Comment (GeorgeMac) [reply internal/storage/storage.go:14]: Yeah, good shout. 25 it is 👍

**Review by markphelps — APPROVED** [review]
🔑
